File: app/src/main/java/com/asf/wallet/interact/ExportWalletInteract.java
package com.asf.wallet.interact;

import com.asf.wallet.entity.Wallet;
import com.asf.wallet.repository.PasswordStore;
import com.asf.wallet.repository.WalletRepositoryType;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;

public class ExportWalletInteract {

  private final WalletRepositoryType walletRepository;
  private final PasswordStore passwordStore;

  public ExportWalletInteract(WalletRepositoryType walletRepository, PasswordStore passwordStore) {
    this.walletRepository = walletRepository;
    this.passwordStore = passwordStore;
  }

  public Single<String> export(Wallet wallet, String backupPassword) {
    return passwordStore.getPassword(wallet)
        .flatMap(password -> walletRepository.exportWallet(wallet, password, backupPassword))
        .observeOn(AndroidSchedulers.mainThread());
  }
}

File: app/src/main/java/com/asfoundation/wallet/interact/ExportWalletInteract.kt
package com.asfoundation.wallet.interact

import com.asfoundation.wallet.entity.Wallet
import com.asfoundation.wallet.repository.PasswordStore
import com.asfoundation.wallet.repository.WalletRepositoryType
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers

class ExportWalletInteract(private val walletRepository: WalletRepositoryType,
                           private val passwordStore: PasswordStore) {

  fun export(wallet: Wallet, backupPassword: String?): Single<String> {
    return passwordStore.getPassword(wallet.address)
        .flatMap { walletRepository.exportWallet(wallet, it, backupPassword) }
        .observeOn(AndroidSchedulers.mainThread())
  }

}